_posts/2023-03-12-gs_00008_programming_II.md

```diff
@@ -0,0 +1,106 @@
+---
+layout: single
+title: Programming Tutorial II
+category: Getting Started
+permalink: /categories/getting_started/programming_tutorial_II
+excerpt: "Basic programming concepts."
+description:
+  Second part of our programming tutorial blog series. Basic programming
+  concepts.
+author: Bjoern Hartmann
+toc: true
+toc_label: "Content"
+toc_sticky: true
+toc_icon: "align-justify"
+sidebar:
+  nav: "getting_started_0"
+header:
+  overlay_color: "#000"
+  overlay_filter: "0.5"
+  overlay_image: /assets/images/getting_started.jpeg
+---
+
+Hello programming newbies! Welcome back to the second part of our programming
+tutorial. Last time we explained why we need programming languages. This time we
+want to take an example from our daily life to discuss a few general programming
+concepts. Imagine you wake up Sunday morning and want to make some pancakes. A
+pretty simple recipe could be: "Mix 200g flour, 2 eggs, and 200ml milk. Then,
+bake it in a pan". Let's see what this has to do with programming.
+
+## Pancakes for a programmer
+
+First, it should be clear that you have to follow the recipe **sequentially**.
+You cannot put anything into the pan before you have mixed the ingredients. The
+same holds for programs. They run sequentially. There are also programs running
+in parallel. Imagine a friend helps you with the pancake. You can measure 200ml
+milk and 200g flour in parallel, but as a whole, you still need to follow the
+sequential order (the pan still needs to wait). Parallel programs are similar,
+but here we want to assume pure sequential programs.
+
+Let's start with the eggs. We have to take an egg, break it, and put it into a
+bowl. Also, we need to throw the eggshell into the trash. For the second egg, we
+have to do exactly the same. In a programming language, you can define a
+**function** _put_egg_in_bowl_ to define a sequence that is used multiple times.
+Next, you can **call** this function in a **loop** for every egg. So, a program
+could look like this:
+
+    function put_egg_in_bowl:
+        take egg
+        break egg
+        put it into bowl
+        throw shell into trash
+
+    for 2 times:
+        put_egg_in_bowl
+
+Note, that we use indentation (here we use 4 spaces) to show which instructions
+belong to our function and loop. After the eggs, we want to weigh the flour. How
+do you usually do it? You watch the scale. Until your target weight of 200g is
+not reached you put more flour into the bowl. In programming languages, there
+exist two types of loops. For the one we used above, we know how often we need
+to repeat it. This is called a **for-loop**. For weighing the flour, this is not
+very useful since you don't know how often you have to put more. We need a loop
+that continues while a condition is true. This is called **while-loop**:
+
+    while flour < 200g:
+        put more flour
+
```

Comment:
You can read it like this: "Repeat putting more flour _while_ the weight is less than 200g." Once the weight of the flour reaches 200g, the while-loop is done.